cmd/cloudFoundryDeploy.go

```diff
@@ -0,0 +1,855 @@
+			log.Entry().Warningf("Blue green deployment is not possible for application without route. Using deployment type '%s' instead.", deployTypeStandard)
+			return deployTypeStandard, nil
+		}
+	}
+
+	return config.DeployType, nil
+}
+
+func deployMta(config *cloudFoundryDeployOptions, mtarFilePath string, command command.ExecRunner) error {
+
+	deployCommand := "deploy"
+	deployParams := []string{}
+
+	if len(config.MtaDeployParameters) > 0 {
+		deployParams = append(deployParams, strings.Split(config.MtaDeployParameters, " ")...)
+	}
+
+	if config.DeployType == "bg-deploy" || config.DeployType == "blue-green" {
+
+		deployCommand = "bg-deploy"
+
+		const noConfirmFlag = "--no-confirm"
+		if !contains(deployParams, noConfirmFlag) {
+			deployParams = append(deployParams, noConfirmFlag)
+		}
+	}
+
+	cfDeployParams := []string{
+		deployCommand,
+		mtarFilePath,
+	}
+
+	if len(deployParams) > 0 {
+		cfDeployParams = append(cfDeployParams, deployParams...)
+	}
+
+	cfDeployParams = append(cfDeployParams, handleMtaExtensionDescriptors(config.MtaExtensionDescriptor)...)
+
+	return cfDeploy(config, cfDeployParams, nil, nil, command)
+}
+
+func handleMtaExtensionDescriptors(mtaExtensionDescriptor string) []string {
+	var result = []string{}
+	for _, part := range splitAtWhitespace(strings.Trim(mtaExtensionDescriptor, " ")) {
+		if part == "-e" || part == "" {
+			continue
+		}
+		// REVISIT: maybe check if the extension descriptor exists
+		result = append(result, "-e", part)
+	}
+	return result
+}
+
+// would make sense to have that method in some kind of helper instead having it here
+func contains(collection []string, key string) bool {
+
+	for _, v := range collection {
+		if v == key {
+			return true
+		}
+	}
+	return false
+}
+
+func cfDeploy(
+	config *cloudFoundryDeployOptions,
+	cfDeployParams []string,
+	additionalEnvironment []string,
+	postDeployAction func(command command.ExecRunner) error,
+	command command.ExecRunner) error {
+
+	const cfLogFile = "cf.log"
+	var err error
+	var loginPerformed bool
+
+	if additionalEnvironment == nil {
+		additionalEnvironment = []string{}
+	}
+	additionalEnvironment = append(additionalEnvironment, "CF_TRACE="+cfLogFile)
+
+	if len(config.CfHome) > 0 {
+		additionalEnvironment = append(additionalEnvironment, "CF_HOME="+config.CfHome)
+	}
+
+	if len(config.CfPluginHome) > 0 {
+		additionalEnvironment = append(additionalEnvironment, "CF_PLUGIN_HOME="+config.CfPluginHome)
+	}
+
+	log.Entry().Infof("Using additional environment variables: %s", additionalEnvironment)
+
+	// TODO set HOME to config.DockerWorkspace
+	command.SetEnv(additionalEnvironment)
+
+	err = _cfLogin(command, cloudfoundry.LoginOptions{
+		CfAPIEndpoint: config.APIEndpoint,
+		CfOrg:         config.Org,
+		CfSpace:       config.Space,
+		Username:      config.Username,
+		Password:      config.Password,
+		CfAPIOpts:     splitAtWhitespace(config.APIParameters),
+		CfLoginOpts:   splitAtWhitespace(config.LoginParameters),
```

Comment:
I know this may be a breaking change compared to the Groovy step, but if these parameters could in theory contain spaces, then splitting them like that does not work. (Because they cannot be escaped at the same time.)

Reply:
For `login` we have: `[--sso | --sso-passcode PASSCODE] [--origin ORIGIN]`
For `api`we have: `--skip-ssl-validation --unset` (both without value``

Here in this case I bet that we won't see blanks here (I'm aware that the time gap between "this will not happen" and the point in time where it happens is sometimes small ...). The passcode does (apparently) not contain blanks (at the moment, I tried it), the examples I find for origin suggests the same (I know, I know, this is a weak answer).

The funny thing is: we do not have a `cf api` call when we login to cf. Hence that CfAPIOpts needs to be removed. -> Will adjust the CfAPIOpts in another PR today.

Reply:
Ok.